src/peft/tuners/delora/layer.py

```diff
@@ -0,0 +1,630 @@
+        # Misc flag
+        "use_residual_init",
+    )
+
+    def __init__(self, base_layer: nn.Module, **kwargs) -> None:
+        self.base_layer = base_layer
+        self.r = {}
+        self.delora_dropout = nn.ModuleDict({})
+        self.delora_A = nn.ParameterDict({})
+        self.delora_B = nn.ParameterDict({})
+        self.delora_alpha = nn.ParameterDict({})
+        # Store flag from kwargs (defaults to True if not provided)
+        self.use_residual_init = bool(kwargs.get("use_residual_init", True))
+        # Non-trainable initial copies so that delta is zero at init
+        # Use persistent buffers so they are included in state_dict and saved.
+        self.delora_initial_A = BufferDict({}, persistent=True)
+        self.delora_initial_B = BufferDict({}, persistent=True)
+        # store alpha as tensor buffer (shape [1]) for easier device moves and persistence
+        self.delora_initial_alpha = BufferDict({}, persistent=True)
+        self.delora_initial_w_norm = BufferDict({}, persistent=True)  # initial base weight norm per adapter
+        self.delora_w_norm = BufferDict({}, persistent=True)  # fixed reference norm per adapter used for delta computation
+        # Track which adapters had their initial delta subtracted from base weights
+        self._initials_applied = set()
+        self._just_loaded = False  # Flag to prevent double-correction on unmerge after load
+        # Track nesting depth of disable->enable toggling to avoid double-adding/subtracting
+        self._disable_depth = 0
+        # Mark the weight as unmerged
+        self._disable_adapters = False
+        self.merged_adapters = []
+        # No merged delta caching; exact unmerge relies on fixed per-adapter reference norm
+        self.kwargs = kwargs
+
+        base_layer_mod = self.get_base_layer()
+        if isinstance(base_layer_mod, nn.Linear):
+            self.in_features, self.out_features = base_layer_mod.in_features, base_layer_mod.out_features
+        else:
+            raise ValueError(f"Unsupported layer type {type(base_layer_mod)}")
+
+    @staticmethod
+    def _compute_delta_from_params(
+        A: torch.Tensor, B: torch.Tensor, alpha: torch.Tensor, r: int, w_norm: torch.Tensor
+    ) -> torch.Tensor:
+        """Compute delta = B @ diag(alpha/r / (||A_i||*||B^j||)) @ A, scaled by provided w_norm (per-input channel)"""
+        comp_dtype = torch.float32 if A.dtype in (torch.float16, torch.bfloat16) else A.dtype
+        A_ = A.to(comp_dtype)
+        B_ = B.to(comp_dtype)
+        alpha_ = alpha.to(comp_dtype)
+        w_norm_ = w_norm.to(comp_dtype)
+        An = torch.clamp(A_.norm(dim=1), min=1e-12)
+        Bn = torch.clamp(B_.norm(dim=0), min=1e-12)
+        diag = torch.diag_embed((alpha_ / r) / (An * Bn))
+        delta = B_ @ diag @ A_
+        delta = delta * w_norm_.unsqueeze(0)
+        return delta.to(A.dtype, copy=False)
+
+    def get_delta_weight(self, adapter: str) -> torch.Tensor:
```

Comment:
There is a lot of similarity between this method and `_compute_delta_from_params`, I wonder if, we a little refactoring of `_compute_delta_from_params`, it could be reused here. WDYT?

Reply:
Thanks for the suggestion, I refactored the code to reuse the same `_compute_delta` function